README.md

# sententia <img src="man/figures/logo.png" align="right" alt="" width="120" />

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![](https://img.shields.io/github/last-commit/71point4/sententia.svg)](https://github.com/71point4/sententia/commits/develop)

[Sententia](https://github.com/71point4/sententia) provides access to
media sentiment data from the Bureau for Economic research.

This is the homepage for the {sententia} R package is at
<https://github.com/71point4/sententia>.

The homepage for the {sententia} `python` package is at
<https://github.com/71point4/pysententia>.

## Installation

Install from GitHub.

``` r
remotes::install_github("71point4/sententia")
```

If you want availability to the latest features (at your own risk) then
you can also install from the development branch.

``` r
remotes::install_github("71point4/sententia", ref = "dev")
```

## Usage

``` r
library(sententia)
```

Check version.

``` r
packageVersion("sententia")
```

### Set API Key

To access the API you’ll need to first specify an API key as provided to
you by the BER.

``` r
# Example API key (this key will not work).
set_api_key("7b58ffa5afcf893d678433a56e0edca5")
get_api_key()
```

If you wish not to set the key every time, use
`usethis::edit_r_environ()` to set the key to `SENTENTIA_KEY`.

Please be aware, that there is an order where the API looks for keys:

1.  Look for `ENV` variable, `SENTENTIA_KEY`, if not found, look in
    `cache` where key was stored using `set_api_key()`
2.  `set_api_key()` OVERRIDES `SENTENTIA_KEY` in `.Renviron`!

To obtain a key, please get in touch. Contact details are in
`DESCRIPTION`.

# The API interface

Besides providing sentiment calculations from different word-list
dictionaries, the API interface provides access to the various
permutations that is available in calculating a sentiment score (Within
text and across time):

<img src="man/figures/aggregations.png" width="1236" style="display: block; margin: auto;" />

-   Sentiment calculation `WITHIN` the article
    -   In the API this is set by the `aggr` parameter
-   Sentiment calculation `ACCROSS` a time period
    -   This does not need to be set. API returns four aggregations:
        -   `mean_sentiment`
        -   `relative`
        -   `absolute`
        -   `sent_log`

## Sentiment Index

Get media sentiment index values for specified media source, model,
topic, dictionary, frequency, and aggregation method combination.

``` r
sent_index(
   source = "businessday",
   model = "model_2021-05-15",
   topic = "global",
   freq = "day",
   dict = "loughran",
   aggr = "sent_logit"
   )

# # A tibble: 224 × 9
#    source      published_date dict_name type     aggr_type  mean_sentiment relative absolute sent_log
#    <chr>       <chr>          <chr>     <chr>    <chr>               <dbl>    <dbl>    <dbl>    <dbl>
#  1 businessday 2003-01-01     loughran  polarity sent_logit         -1.80    -1       -1        0    
#  2 businessday 2003-02-01     loughran  polarity sent_logit         -0.784   -0.639   -0.576   -0.645
#  3 businessday 2003-03-01     loughran  polarity sent_logit         -0.768   -0.539   -0.505   -0.523
#  4 businessday 2003-04-01     loughran  polarity sent_logit         -0.604   -0.423   -0.420   -0.388
#  5 businessday 2003-05-01     loughran  polarity sent_logit         -0.845   -0.582   -0.542   -0.567
#  6 businessday 2003-06-01     loughran  polarity sent_logit         -0.594   -0.426   -0.390   -0.394
#  7 businessday 2003-07-01     loughran  polarity sent_logit         -0.672   -0.462   -0.439   -0.433
#  8 businessday 2003-08-01     loughran  polarity sent_logit         -0.502   -0.356   -0.337   -0.323
#  9 businessday 2003-09-01     loughran  polarity sent_logit         -0.637   -0.478   -0.435   -0.452
# 10 businessday 2003-10-01     loughran  polarity sent_logit         -0.604   -0.412   -0.384   -0.381
# # … with 214 more rows
```

# Count of articles that make up sentiment

Get a count of the number of articles for a specified media source,
model, topic, and frequency of aggregation.

``` r
sent_counts(
   source = "all",
   model = "model_2021-05-15",
   topic = "global",
   freq = "week"
   )

# # A tibble: 972 × 2
#    published_date article_count
#    <chr>                  <int>
#  1 2003-01-13                 1
#  2 2003-01-20                 1
#  3 2003-01-27                 1
#  4 2003-02-03                 3
#  5 2003-02-10                 7
#  6 2003-02-17                26
#  7 2003-02-24                56
#  8 2003-03-03                90
#  9 2003-03-10               171
# 10 2003-03-17               152
# # … with 962 more rows
```

# Date polarity

Get a count of the number of positive and negative articles for a
specified model, topic, dictionary, aggregation method, and frequency.

``` r
sent_date_polarity(
   source = "all",
   model = "model_2021-05-15",
   topic = "global",
   freq = "week",
   dict = "loughran",
   aggr = "sent_logit"
   )

# # A tibble: 972 × 7
#    source published_date dict_name type     aggr_type  positive_words negative_words
#    <chr>  <chr>          <chr>     <chr>    <chr>               <int>          <int>
#  1 all    2003-01-13     loughran  polarity sent_logit              0              1
#  2 all    2003-01-20     loughran  polarity sent_logit              0              1
#  3 all    2003-01-27     loughran  polarity sent_logit              0              1
#  4 all    2003-02-03     loughran  polarity sent_logit              0              3
#  5 all    2003-02-10     loughran  polarity sent_logit              0              7
#  6 all    2003-02-17     loughran  polarity sent_logit              1             25
#  7 all    2003-02-24     loughran  polarity sent_logit             14             33
#  8 all    2003-03-03     loughran  polarity sent_logit             20             67
#  9 all    2003-03-10     loughran  polarity sent_logit             31            130
# 10 all    2003-03-17     loughran  polarity sent_logit             39            105
# # … with 962 more rows
```

# Word polarity

Get the top 50 most frequently occurring positive and negative words for
a specified model, topic, dictionary, aggregation method, and frequency.
The timeframe over which these words are selected depends on the
specified frequency (day = 30 days, week = 3 months, month = 6 months).

``` r
sent_word_polarity(
   source = "all",
   model = "model_2021-05-15",
   topic = "economy",
   freq = "day",
   dict = "loughran"
   )

# # A tibble: 100 × 8
#    source topic_econ tokens       dict_name sentiment mentions word_rank date_range             
#    <chr>  <chr>      <chr>        <chr>     <chr>        <int>     <int> <chr>                  
#  1 all    economy    against      loughran  negative      1681         1 2021-02-28 - 2021-08-29
#  2 all    economy    concerns     loughran  negative       856         2 2021-02-28 - 2021-08-29
#  3 all    economy    lost         loughran  negative       797         3 2021-02-28 - 2021-08-29
#  4 all    economy    crisis       loughran  negative       754         4 2021-02-28 - 2021-08-29
#  5 all    economy    losses       loughran  negative       608         5 2021-02-28 - 2021-08-29
#  6 all    economy    unrest       loughran  negative       586         6 2021-02-28 - 2021-08-29
#  7 all    economy    unemployment loughran  negative       562         7 2021-02-28 - 2021-08-29
#  8 all    economy    challenges   loughran  negative       554         8 2021-02-28 - 2021-08-29
#  9 all    economy    late         loughran  negative       534         9 2021-02-28 - 2021-08-29
# 10 all    economy    violence     loughran  negative       530        10 2021-02-28 - 2021-08-29
# # … with 90 more rows
```


NEWS.md
# sententia 0.1.2

* Removal of parameter checking

# sententia 0.1.0

* Launch of the basic API.


R/utils.R
#' Check that API response is JSON
#'
#' @param response A response object.
#'
#' @return NULL. Raises an error if the response is not JSON.
check_response_json <- function(response) {
  if (http_type(response) != "application/json") {
    stop("API did not return json.", call. = FALSE)
  }
}

#' Check API response for errors
#'
#' @param response A response object.
#'
#' @return NULL. Raises an error if response has an error code.
check_response_error <- function(response) {
  if (http_error(response)) {
    status <- status_code(response)
    error <- glue("API request failed [{status}]")
    
    message <- content(response)$message
    if (!is.null(message)) {
      error <- glue("{error}: {message}")
    }
    
    stop(error, call. = FALSE
    )
  }
}

#' Check if user has specified all the parameters
#'
#' @return character
missing_args <- function()
{
  calling_function <- sys.function(1)
  all_args <- names(formals(calling_function))
  all_args <- all_args[all_args != "..."]
  matched_call <- match.call(
    calling_function, 
    sys.call(1), 
    expand.dots = FALSE
  )
  passed_args <- names(as.list(matched_call)[-1])
  setdiff(all_args, passed_args)
}

#' Wrapper for httr::GET()
#'
#' @param url URL to retrieve.
#' @param config Additional configuration settings.
#' @param ... Further named parameters.
#' @param retry Number of times to retry request on failure.
GET <- function(url = NULL, config = list(), retry = 5, ...) {
  headers = list()
  
  api_key <- NULL
  try(api_key <- get_api_key(), silent = TRUE)
  
  if (!is.null(api_key)) {
    headers = list("token" = api_key)
  }

  response <- httr::RETRY(
    "GET",
    url,
    config,
    httr::user_agent(glue("sententia [R] ({PKG_VERSION})")),
    ...,
    do.call(add_headers, headers),
    handle = NULL,
    times = retry,
    terminate_on = c(400, 404, 429, 500)
  )
  
  # Check for "400 UNAUTHORISED".
  # Check for "429 LIMIT EXCEEDED".
  #
  if (response$status_code %in% c(400, 429)) {
      out <- response %>% 
        content(as = "text", encoding = "UTF-8") %>%
        fromJSON() %>%
        toJSON(pretty = TRUE)
      
      stop(out, call. = FALSE)
  }
  
  response
}


R/sententia.R
BASE_URL <- "https://berapi.71point4.com/v1/"

PKG_VERSION <- utils::packageDescription('sententia')$Version

cache <- new.env()

# Avoid some "no visible binding" notes with devtools::check().
# globalVariables(
#   c(
#   )
# )

#' @import jsonlite
#' @import tibble
#' @importFrom dplyr arrange
#' @importFrom magrittr %>%
#' @importFrom utils URLencode
#' @importFrom glue glue
#' @importFrom glue glue_collapse
#' @importFrom httr add_headers
#' @importFrom httr http_error
#' @importFrom httr http_type
#' @importFrom httr content
#' @importFrom httr status_code

assign("base_url", BASE_URL, envir = cache)

#' Set IP and port for API server when testing
#'
#' @param ip IP address for API server.
#' @param port Port on which API is running.
#'
#' @export
#'
#' @examples
#' # Use local server
#' set_server("0.0.0.0", 8080)
set_server <- function(ip, port) {
  assign("base_url", sprintf("http://%s:%d/", ip, port), envir = cache)
}

#' Set base URL for API  when testing
#'
#' @param url URL to access API.
#'
#' @export
#'
#' @examples
#' # Use local server
#' set_base_url("http://0.0.0.0:8080/")
set_base_url <- function(url) {
  assign("base_url", url, envir = cache)
}

#' Get base URL for API
#'
#' @return API base URL.
base_url <- function() {
  get("base_url", envir = cache)
}

#' Set API key
#'
#' Specify the key for accessing the sententia API. 
#' Automatically determines which type of key is being used and authenticates appropriately.
#'
#' @param api_key API key.
#'
#' @export
#'
#' @examples
#' # Authenticate using an API key.
#' set_api_key("412dd2813fbf9037f5266557bbf5d1f5")
#'
set_api_key <- function(api_key) {

  if (nchar(api_key) == 32) {
    if(Sys.getenv("SENTENTIA_KEY") != ""){
      message("WARNING, an API key was found in .Renviron: SENTENTIA_KEY, it was replaced with key supplied.")  
      assign("api_key", api_key, envir = cache)
    } else {
      assign("api_key", api_key, envir = cache)
    }
  } else {
    if (api_key == "" || is.na(api_key) || is.null(api_key)) {
      stop("API key is missing.", call. = FALSE)
    } else {
      stop(glue("Unknown API key type: '{api_key}'."), call. = FALSE)
    }
  }
}

#' Retrieve API key
#'
#' @return API key.
#' @export
#'
#' @examples
#' \dontrun{
#' set_api_key("412dd2813fbf9037f5266557bbf5d1f5")
#' get_api_key()
#' }
get_api_key <- function() {
  if(Sys.getenv("SENTENTIA_KEY") == ""){
    api_key <- try(get("api_key", envir = cache), silent = TRUE)
  }
  
  if(Sys.getenv("SENTENTIA_KEY") != ""){
    api_key <- try(get("api_key", envir = cache), silent = TRUE)
    if (class(api_key) == "try-error" || api_key == ""){
      api_key <- Sys.getenv("SENTENTIA_KEY")
    } 
  }
  
  if (class(api_key) == "try-error" || api_key == "") {
    stop("Use set_api_key() to specify an API key.", call. = FALSE)
  } else {
    api_key
  }
}


R/sent_count.R
#' Get a count of the number of articles for a specified media source.
#'
#' @param source The news source for which data is sought. Data can either be drawn from all sources (sourc=all), from News24 only (source=news24), or from Business Day only (source=businessday).. 
#' @param model The name of the model to use. Even the general index (i.e. no topics) requires a model name.
#' @param topic The topic which you are interested in.
#' @param freq 	The frequency on which to aggregate the index. Available values: day, week, month.
#'
#' @return tibble.
#' @export
#'
#' @examples
#' # Get a count of the number of articles for a specified media source, model, topic, and frequency of aggregation.
#' \dontrun{
#' sent_counts(
#'    source="all",
#'    model="model_2021-05-15",
#'    topic="global",
#'    freq="month",
#'    )
#' }
#' 
sent_counts <- function(source, model, topic, freq){
  
  url <- glue(base_url(), "counts?source={source}&model={model}&topic={topic}&freq={freq}")
  
  response <- GET(url)
  
  check_response_error(response)
  check_response_json(response)
  
  result <- response %>%
    content(as = "text", encoding = "UTF-8") %>%
    fromJSON() %>%
    arrange(published_date) %>% 
    as_tibble()
  
  return(result)
  
}

R/sent_index.R
#' Pull a specific media sentiment index
#'
#' @param source The news source for which data is sought. Data can either be drawn from all sources (sourc=all), from News24 only (source=news24), or from Business Day only (source=businessday).. 
#' @param model The name of the model to use. Even the general index (i.e. no topics) requires a model name.
#' @param topic The topic which you are interested in.
#' @param freq 	The frequency on which to aggregate the index. Available values: day, week, month.
#' @param dict The dictionary to use in sentiment characterization for the index. Available values: afinn, anew, geninqposneg, huliu, loughran, lsd2015, nrc, vader.
#' @param aggr The within article aggregation method to use. Available values: sent_abspropdiff, sent_relpropdiff, sent_logit, sent_vader, sent_valence.
#'
#' @return tibble.
#' @export
#'
#' @examples
#' # Get global sentiment index.
#' \dontrun{
#' sent_index(
#'    source="all",
#'    model="model_2021-05-15",
#'    topic="global",
#'    freq="month",
#'    dict="loughran",
#'    aggr="sent_logit"
#'    )
#' }
#' 
sent_index <- function(source, model, topic, freq, dict, aggr){

  url <- glue(base_url(), "index?source={source}&model={model}&topic={topic}&freq={freq}&dict={dict}&aggr={aggr}")
  
  response <- GET(url)
  
  check_response_error(response)
  check_response_json(response)
  
  result <- response %>%
    content(as = "text", encoding = "UTF-8") %>%
    fromJSON() %>%
    arrange(published_date) %>% 
    as_tibble()
  
  return(result)
  
}

R/sent_date_polarity.R
#' Get a count of the number of positive and negative articles.
#'
#' @param source The news source for which data is sought. Data can either be drawn from all sources (sourc=all), from News24 only (source=news24), or from Business Day only (source=businessday).. 
#' @param model The name of the model to use. Even the general index (i.e. no topics) requires a model name.
#' @param topic The topic which you are interested in.
#' @param freq 	The frequency on which to aggregate the index. Available values: day, week, month.
#' @param dict The dictionary to use in sentiment characterization for the index. Available values: afinn, anew, geninqposneg, huliu, loughran, lsd2015, nrc, vader.
#' @param aggr The aggregation method to use. Available values: sent_abspropdiff, sent_relpropdiff, sent_logit, sent_vader, sent_valence.
#'
#' @return tibble.
#' @export
#'
#' @examples
#' # Get a count of the number of positive and negative articles for a specified model, topic, dictionary, aggregation method, and frequency.
#' \dontrun{
#' sent_date_polarity(
#'    source="all",
#'    model="model_2021-05-15",
#'    topic="global",
#'    freq="month",
#'    dict="loughran",
#'    aggr="sent_logit"
#'    )
#' }
#' 
sent_date_polarity <- function(source, model, topic, freq, dict, aggr){

  url <- glue(base_url(), "date_polarity?source={source}&model={model}&topic={topic}&freq={freq}&dict={dict}&aggr={aggr}")
  
  response <- GET(url)
  
  check_response_error(response)
  check_response_json(response)
  
  result <- response %>%
    content(as = "text", encoding = "UTF-8") %>%
    fromJSON() %>%
    arrange(published_date) %>% 
    as_tibble()
  
  return(result)
  
}

R/sent_word_polarity.R
#' Get the top 50 most frequently occurring positive and negative words.
#' @description Get the top 50 most frequently occurring positive and negative words for a specified model, topic, dictionary, aggregation method, and frequency. The timeframe over which these words are selected depends on the specified frequency (day = 30 days, week = 3 months, month = 6 months).
#' @param source The news source for which data is sought. Data can either be drawn from all sources (sourc=all), from News24 only (source=news24), or from Business Day only (source=businessday).. 
#' @param model The name of the model to use. Even the general index (i.e. no topics) requires a model name.
#' @param topic The topic which you are interested in.
#' @param freq 	The frequency on which to aggregate the index. Available values: day, week, month.
#' @param dict The dictionary to use in sentiment characterization for the index. Available values: afinn, anew, geninqposneg, huliu, loughran, lsd2015, nrc, vader.
#'
#' @return tibble.
#' @export
#'
#' @examples
# Get the top 50 most frequently occurring positive and negative words. (day = 30 days, week = 3 months, month = 6 months)
#' 
#' \dontrun{
#' sent_word_polarity(
#'    source="all",
#'    model="model_2021-05-15",
#'    topic="global",
#'    freq="month",
#'    dict="loughran"
#'    )
#' }
#' 
sent_word_polarity <- function(source, model, topic, freq, dict){
  
  url <- glue(base_url(), "word_polarity?source={source}&model={model}&topic={topic}&freq={freq}&dict={dict}")
  
  response <- GET(url)
  
  check_response_error(response)
  check_response_json(response)
  
  result <- response %>%
    content(as = "text", encoding = "UTF-8") %>%
    fromJSON() %>%
    as_tibble()
  
  return(result)
  
}